src/System.IO.FileSystem.AccessControl/tests/DirectoryObjectSecurityTests.cs

```diff
@@ -118,26 +285,266 @@ public void AddAuditRule_InvalidObjectAuditRule()
+        }
+
+        [Fact]
         public void RemoveAccessRuleSpecific_InvalidObjectAccessRule()
         {
             var customObjectSecurity = new CustomDirectoryObjectSecurity();
             AssertExtensions.Throws<ArgumentNullException>("rule", () => customObjectSecurity.RemoveAccessRuleSpecific(null));
         }
 
         [Fact]
+        public void RemoveAccessRuleSpecific_AccessControlType_Allow_Succeeds()
+        {
+            var descriptor = new CommonSecurityDescriptor(true, true, string.Empty);
+            var customObjectSecurity = new CustomDirectoryObjectSecurity(descriptor);
+
+            // read access mask  : 0x01
+            // write access mask : 0x02
+            int readWriteAccessMask = 0x01 | 0x02;
+            var objectTypeGuid = Guid.NewGuid();
+            var identityReference = new NTAccount(@"NT AUTHORITY\SYSTEM");
+
+            var customAccessRuleReadWrite = new CustomAccessRule(
+                identityReference, readWriteAccessMask, true, InheritanceFlags.None,
+                PropagationFlags.None, objectTypeGuid, Guid.NewGuid(), AccessControlType.Allow
+                );
+
+            customObjectSecurity.AddAccessRule(customAccessRuleReadWrite);
+            customObjectSecurity.RemoveAccessRuleSpecific(customAccessRuleReadWrite);
+
+            AuthorizationRuleCollection ruleCollection =
+                customObjectSecurity
+               .GetAccessRules(true, true, typeof(System.Security.Principal.NTAccount));
+
+            List<CustomAccessRule> existingRules = ruleCollection.Cast<CustomAccessRule>().ToList();
+            Assert.False(existingRules.Contains(customAccessRuleReadWrite));
+        }
+
+        [Fact]
+        public void RemoveAccessRuleSpecific_AccessControlType_Deny_NoMatchableRules_Succeeds()
+        {
+            var descriptor = new CommonSecurityDescriptor(true, true, string.Empty);
+            var customObjectSecurity = new CustomDirectoryObjectSecurity(descriptor);
+
+            // read access mask  : 0x01
+            // write access mask : 0x02
+            int readWriteAccessMask = 0x01 | 0x02;
+            var objectTypeGuid = Guid.NewGuid();
+            var identityReference = new NTAccount(@"NT AUTHORITY\SYSTEM");
+
+            var customAccessRuleReadWrite = new CustomAccessRule(
+                identityReference, readWriteAccessMask, true, InheritanceFlags.None,
+                PropagationFlags.None, objectTypeGuid, Guid.NewGuid(), AccessControlType.Deny
+                );
+
+            int readAccessMask = 0x01;
+            var customAccessRuleRead = new CustomAccessRule(
+                new NTAccount(@"NT AUTHORITY\SYSTEM"), readAccessMask, true, InheritanceFlags.None,
+                PropagationFlags.None, objectTypeGuid, Guid.NewGuid(), AccessControlType.Deny
+                );
+
+            customObjectSecurity.AddAccessRule(customAccessRuleReadWrite);
+            customObjectSecurity.RemoveAccessRuleSpecific(customAccessRuleRead);
```

Comment:
You have a point here but I thought it is maybe more readable to have all the code in one place ( in case of unit tests ) for those who just want to read and understand the test case.